src/plugins/fs/exfat.c

```diff
@@ -0,0 +1,302 @@
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether the exfat file system on the @device was successfully repaired
+ *          (if needed) or not (error is set in that case)
+ *
+ * Tech category: %BD_FS_TECH_EXFAT-%BD_FS_TECH_MODE_REPAIR
+ */
+gboolean bd_fs_exfat_repair (const gchar *device, const BDExtraArg **extra, GError **error) {
+    const gchar *args[4] = {"exfatfsck", "-a", device, NULL};
+    gint status = 0;
+    gboolean ret = FALSE;
+
+    if (!check_deps (&avail_deps, DEPS_EXFATFSCK_MASK, deps, DEPS_LAST, &deps_check_lock, error))
+        return FALSE;
+
+    ret = bd_utils_exec_and_report_status_error (args, extra, &status, error);
+    if (!ret && (status == 1)) {
+        /* no error should be reported for exit code 1 */
+        g_clear_error (error);
+        ret = TRUE;
+    }
+
+    return ret;
+}
+
+/**
+ * bd_fs_exfat_set_label:
+ * @device: the device containing the file system to set label for
+ * @label: label to set
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether the label of exfat file system on the @device was
+ *          successfully set or not
+ *
+ * Tech category: %BD_FS_TECH_EXFAT-%BD_FS_TECH_MODE_SET_LABEL
+ */
+gboolean bd_fs_exfat_set_label (const gchar *device, const gchar *label, GError **error) {
+    const gchar *args[5] = {"exfatlabel", device, label, NULL};
+
+    if (!check_deps (&avail_deps, DEPS_EXFATLABEL_MASK, deps, DEPS_LAST, &deps_check_lock, error))
+        return FALSE;
+
+    if (strlen (label) > 15)
+        bd_utils_log (BD_UTILS_LOG_WARNING, "Given label is too long for exFAT and will be truncated." \
+                                            "Labels on exFAT can be at most 15 characters long");
+
+    return bd_utils_exec_and_report_error (args, NULL, error);
+}
+
+/**
+ * bd_fs_exfat_get_info:
+ * @device: the device containing the file system to get info for
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: (transfer full): information about the file system on @device or
+ *                           %NULL in case of error
+ *
+ * Tech category: %BD_FS_TECH_EXFAT-%BD_FS_TECH_MODE_QUERY
+ */
+BDFSExfatInfo* bd_fs_exfat_get_info (const gchar *device, GError **error) {
+    const gchar *args[3] = {"dumpexfat", device, NULL};
+    gboolean success = FALSE;
+    gchar *output = NULL;
+    BDFSExfatInfo *ret = NULL;
+    gchar **lines = NULL;
+    gchar **line_p = NULL;
+    gint scanned = 0;
+
+    if (!check_deps (&avail_deps, DEPS_DUMPEXFAT_MASK, deps, DEPS_LAST, &deps_check_lock, error))
+        return FALSE;
+
+    ret = g_new0 (BDFSExfatInfo, 1);
+
+    success = get_uuid_label (device, &(ret->uuid), &(ret->label), error);
+    if (!success) {
+        /* error is already populated */
+        bd_fs_exfat_info_free (ret);
+        return NULL;
+    }
+
+    success = bd_utils_exec_and_capture_output (args, NULL, &output, error);
+    if (!success)
+        /* error is already populated */
+        return FALSE;
```

Comment:
Well, `FALSE` is basically `NULL` in C but I think it's a copy-paste error ;-)

Reply:
Fixed.

Reply:
Thanks for fixing, what about freeing the `ret`?

Reply:
Oops, also fixed.